### ISSUE TITLE
no-contradicting-variants: false positive when twMerge composes classes from a component's own cva variants

### ISSUE THREAD

**Issue by fmhall** [sergioazoc/oxlint-tailwindcss#117]
Thanks for the plugin — the `entryPoint` design and the `designSystemUnavailable` diagnostic are a real improvement over what we were using, and the v4 deprecation rules found genuine issues in our codebase on day one.

One report on `no-contradicting-variants`, because acting on its findings changed rendering for us.

## Summary

In the `cn()`/`twMerge` + `cva` composition that shadcn-style codebases use, a variant-prefixed class that looks redundant inside its own string is often **load-bearing**: it exists so `twMerge` evicts a conflicting class contributed by the component's own variants. Removing it restores the class it was there to defeat.

## Reproduction

Versions: `oxlint-tailwindcss@1.9.0`, `oxlint@1.78.0`, `tailwindcss@4.3.3`, `tailwind-merge@3.6.0`, `class-variance-authority@0.7.1`, Node 24.13.0

`src/theme.css`
```css
@import "tailwindcss";
```

`src/button.tsx`
```tsx
import { cva } from "class-variance-authority";
import { twMerge } from "tailwind-merge";

const buttonVariants = cva("rounded", {
  variants: { variant: { ghost: "hover:bg-accent" } },
  defaultVariants: { variant: "ghost" },
});

export function Button({ className }: { className?: string }) {
  return <button className={twMerge(buttonVariants(), className)} />;
}
```

`src/usage.tsx`
```tsx
import { Button } from "./button";

export const Example = () => (
  <Button className="bg-transparent hover:bg-transparent" />
);
```

`.oxlintrc.json`
```json
{
  "plugins": ["react"],
  "jsPlugins": ["oxlint-tailwindcss"],
  "settings": { "tailwindcss": { "entryPoint": "src/theme.css" } },
  "rules": { "tailwindcss/no-contradicting-variants": "error" }
}
```

Reported:

```
src/usage.tsx:6:21: error tailwindcss(no-contradicting-variants):
  "hover:bg-transparent" is redundant because "bg-transparent" already applies unconditionally.
```

## Why it is a false positive

`tailwind-merge` resolving the actual runtime value:

```js
twMerge("rounded hover:bg-accent", "bg-transparent hover:bg-transparent")
// -> "rounded bg-transparent hover:bg-transparent"      as written

twMerge("rounded hover:bg-accent", "bg-transparent")
// -> "rounded hover:bg-accent bg-transparent"           after removing the class
```

So the button is transparent on hover as written, and gains an accent hover once the "redundant" class is removed.

The analysis is correct for the string in isolation — read alone, `hover:bg-transparent` adds nothing next to `bg-transparent`. It is the scope that is too narrow: the conflicting class lives in another module, is selected at runtime by `cva`, and is resolved by `twMerge`'s conflict groups.

What makes this one costly in practice is that the finding reads as obviously correct, so it invites a fix. We took the advice on one instance and shipped a dark-mode regression: `dark:bg-transparent` was overriding an input primitive's `dark:bg-input/30`, and removing it gave the field a background in dark mode.

## Suggestion

The plugin already recognises `cn`, `clsx`, `twMerge`, `cva`, `ctl`, `tw` as extractors, so it knows when a string is an argument to a merge-aware call. Skipping `no-contradicting-variants` in that position — or gating it behind an option — would remove this class of false positive, since redundancy genuinely is not decidable there without cross-module analysis and emulation of `twMerge` semantics.

Fully understand if you would rather document the limitation than narrow the rule; the other rules in the correctness group have been solid for us, and this one is the only place we hit composition as a blind spot. We have turned it off in our shared preset for now.

**Comment by sergioazoc** [comment]
Thanks for the report, and for the writeup, your analysis did most of the work here.

You nailed the root cause: the rule read the `className` as the element's final class list, when in
a `cn`/`twMerge` + `cva` setup it's an override fragment. `hover:bg-transparent` isn't redundant,
it's load-bearing, it's there to beat the `hover:bg-accent` that `twMerge` pulls in from the
component's own variants.

Fixed in v1.10.0: extracted strings now carry their origin, so `no-contradicting-variants` only runs
on a literal class list on a native host element, and skips anything inside `cn`/`twMerge`/`cva`, on
a custom component's `className`, or in a variable/tagged template. Same fix for
`no-dark-without-light`, which had the identical blind spot (your `dark:bg-input/30` case). It
trades a few false negatives for killing the false positive, which felt right given how confidently
wrong the finding read.

You can safely turn both back on in your preset. Thanks again 🙏

PD: your twMerge conflict-group breakdown was exactly right, that's why an unscoped base can't evict
a variant-scoped class.
